File: gc-modules/gc-module-system/src/main/java/com/gc/system/controller/SysRoleController.java
package com.gc.system.controller;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.gc.common.auth.annotation.NonUrlCheck;
import com.gc.common.auth.utils.AuthUtils;
import com.gc.common.base.message.Result;
import com.gc.starter.crud.controller.BaseController;
import com.gc.starter.crud.query.PageQueryParameter;
import com.gc.starter.log.annotation.Log;
import com.gc.starter.log.constants.LogType;
import com.gc.system.model.SysRoleFunctionPO;
import com.gc.system.model.SysRolePO;
import com.gc.system.pojo.dto.role.RoleMenuSaveDTO;
import com.gc.system.service.SysRoleFunctionService;
import com.gc.system.service.SysRoleService;
import io.swagger.annotations.ApiOperation;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.io.Serializable;
import java.util.List;
import java.util.stream.Collectors;

/**
 * 角色业务层
 * @author shizhongming
 * 2020/1/24 2:21 下午
 */
@RestController
@RequestMapping("sys/role")
@NonUrlCheck
public class SysRoleController extends BaseController<SysRoleService, SysRolePO> {

    private final SysRoleFunctionService sysRoleFunctionService;

    public SysRoleController(SysRoleFunctionService sysRoleFunctionService) {
        this.sysRoleFunctionService = sysRoleFunctionService;
    }

    @Override
    @PostMapping("save")
    @ApiOperation(value = "添加角色")
    @Log(value = "添加角色", type = LogType.ADD)
    @PreAuthorize("hasPermission('sys:role', 'save')")
    public Result<Boolean> save(@RequestBody SysRolePO model) {
        return Result.success(this.service.saveWithUser(model, AuthUtils.getCurrentUserId()));
    }

    @Override
    @PostMapping("update")
    @ApiOperation(value = "更新角色")
    @Log(value = "更新角色", type = LogType.UPDATE)
    @PreAuthorize("hasPermission('sys:role', 'update')")
    public Result<Boolean> update(@RequestBody SysRolePO model) {
        return super.update(model);
    }

    @Override
    @PostMapping("list")
    @ApiOperation(value = "查询角色列表（支持分页、实体类属性查询）")
    public Result<Object> list(@RequestBody PageQueryParameter<String, Object> parameter) {
        return super.list(parameter);
    }

    @Override
    @ApiOperation(value = "添加修改角色")
    @PostMapping("saveUpdate")
    @Log(value = "添加修改角色", type = LogType.UPDATE)
    @PreAuthorize("hasPermission('sys:role', 'save') or hasPermission('sys:role', 'update')")
    public Result<Boolean> saveUpdate(@RequestBody SysRolePO model) {
        return super.saveUpdate(model);
    }

    @Override
    @PreAuthorize("hasPermission('sys:role', 'delete')")
    @ApiOperation(value = "通过ID批量删除角色")
    @PostMapping("batchDeleteById")
    @Log(value = "通过ID批量删除角色", type = LogType.DELETE)
    public Result<Boolean> batchDeleteById(@RequestBody List<Serializable> idList) {
        return super.batchDeleteById(idList);
    }

    @Override
    @ApiOperation(value = "通过ID查询")
    @PostMapping("getById")
    public Result<SysRolePO> getById(@RequestBody Serializable id) {
        return super.getById(id);
    }

    /**
     * 获取角色对应的功能ID集合
     * @param roleId 角色ID
     * @return 功能ID集合
     */
    @ApiOperation(value = "获取角色对应的功能ID集合")
    @PostMapping("listFunctionId")
    public Result<List<Long>> listFunctionId(@RequestBody Long roleId) {
        return Result.success(
                this.sysRoleFunctionService.list(
                        new QueryWrapper<SysRoleFunctionPO>().lambda()
                        .select(SysRoleFunctionPO :: getFunctionId)
                        .eq(SysRoleFunctionPO :: getRoleId, roleId)
                ).stream().map(SysRoleFunctionPO :: getFunctionId)
                .collect(Collectors.toList())
        );
    }


    @ApiOperation(value = "保存角色功能")
    @PostMapping("saveRoleMenu")
    @Log(value = "保存角色功能", type = LogType.UPDATE)
    @PreAuthorize("hasPermission('sys:role', 'update')")
    public Result<Boolean> saveRoleMenu(@RequestBody @Valid RoleMenuSaveDTO parameter) {
        return Result.success(this.service.saveRoleMenu(parameter));
    }
}

File: smart-module-starter/smart-system/src/main/kotlin/com/smart/system/controller/SysRoleController.kt
package com.smart.system.controller

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.smart.common.log.annotation.Log
import com.smart.common.message.Result
import com.smart.common.model.Tree
import com.smart.starter.crud.constants.CRUDConstants
import com.smart.starter.crud.controller.BaseController
import com.smart.starter.crud.query.PageQueryParameter
import com.smart.starter.crud.utils.MybatisUtil
import com.smart.system.model.SysRoleDO
import com.smart.system.model.SysRoleMenuFunctionDO
import com.smart.system.service.SysRoleMenuFunctionService
import com.smart.system.service.SysRoleService
import com.smart.system.service.SysUserRoleService
import org.apache.shiro.authz.annotation.RequiresPermissions
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.util.StringUtils
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

/**
 *
 * @author ming
 * 2019/6/12 下午3:50
 */
@RestController
@RequestMapping("/sys/role")
class SysRoleController : BaseController<SysRoleService, SysRoleDO>() {

    @Autowired
    private lateinit var userRoleService: SysUserRoleService

    @Autowired
    private lateinit var roleMenuFunctionService: SysRoleMenuFunctionService

    /**
     * 查询所有
     */
    @RequiresPermissions("system:role:query")
    @RequestMapping("/listWithAll")
    fun listWithAll(@RequestBody parameter: PageQueryParameter): Result<Any?> {
        parameter[CRUDConstants.WITH_ALL.name] = true
        return this.list(parameter)
    }

    /**
     * 查询树形结构带有组织结构
     */
    @RequiresPermissions("system:role:query")
    @RequestMapping("/listAllTreeWithOrgan")
    fun listAllTreeWithOrgan(): Result<Tree<Any>?> {
        return try {
            Result.success(this.service.listAllTreeWithOrgan())
        } catch (e: Exception) {
            e.printStackTrace()
            Result.failure(e.message)
        }
    }

    /**
     * 查询角色详情
     */
    @RequiresPermissions("system:role:query")
    @RequestMapping("/queryDetail")
    fun queryDetail(@RequestBody roleDO: SysRoleDO): Result<SysRoleDO?> {
        return try {
            Result.success(this.service.queryDetail(roleDO))
        } catch (e: Exception) {
            e.printStackTrace()
            Result.failure(e.message)
        }
    }

    /**
     * 查询角色对应的ID
     */
    @RequestMapping("/getRoleUserId")
    @RequiresPermissions("system:role:query")
    fun getRoleUserId(@RequestBody role: SysRoleDO): Result<List<String>?> {
        try {
            if (StringUtils.isEmpty(role.roleId)) {
                return Result.success(arrayListOf())
            }
            val userRoleMap = this.userRoleService.queryByRoleIds(arrayListOf(role.roleId!!))
            val userRoleList = userRoleMap[role.roleId!!] ?: return Result.success(arrayListOf())
            val userIdList = mutableListOf<String>()
            userRoleList.forEach {
                if (it.roleId == role.roleId) {
                    userIdList.add(it.userId!!)
                }
            }
            return Result.success(userIdList)
        } catch (e: Exception) {
            e.printStackTrace()
            return Result.failure(e.message)
        }
    }

    /**
     * 更新角色对应的人员信息
     */
    @RequiresPermissions("system:role:updateUser")
    @RequestMapping("/updateUser")
    @Log("更新角色人员信息")
    fun updateUser(@RequestBody parameters: Map<String, List<String>>): Result<Boolean?> {
        try {
            return Result.success(this.service.updateUser(parameters))
        } catch (e: Exception) {
            e.printStackTrace()
            return Result.failure(e.message, false)
        }
    }

    /**
     * 获取角色授权信息
     */
    @RequiresPermissions("system:role:query")
    @PostMapping("/queryAuthentication")
    fun queryAuthentication(@RequestBody roleMenuFunction: SysRoleMenuFunctionDO): Result<List<SysRoleMenuFunctionDO>?> {
        try {
            val wrapper = QueryWrapper<SysRoleMenuFunctionDO>()
            wrapper.eq(MybatisUtil.getDbField(SysRoleMenuFunctionDO :: roleId), roleMenuFunction.roleId)
                    .eq(MybatisUtil.getDbField(SysRoleMenuFunctionDO :: menuConfigId), roleMenuFunction.menuConfigId)
            return Result.success(this.roleMenuFunctionService.list(wrapper))
        } catch (e: Exception) {
            e.printStackTrace()
            return Result.failure(e.message, null)
        }
    }

    /**
     * 角色授权
     */
    @RequiresPermissions("system:role:authorize")
    @PostMapping("/authorize")
    @Log("角色授权")
    fun authorize(@RequestBody list: List<SysRoleMenuFunctionDO>): Result<Boolean?> {
        try {
            return Result.success(this.service.authorize(list))
        } catch (e: Exception) {
            e.printStackTrace()
            return Result.failure(e.message, false)
        }
    }

}